src/backups.py

```diff
@@ -0,0 +1,265 @@
+        metadata = f"""Date Backup Requested: {datetime_backup_requested}
+Model Name: {self.model.name}
+Application Name: {self.model.app.name}
+Unit Name: {self.charm.unit.name}
+Juju Version: {str(juju_version)}
+"""
+        success = upload_content_to_s3(
+            metadata,
+            s3_parameters["bucket"],
+            f"{s3_directory}/metadata",
+            s3_parameters["region"],
+            s3_parameters["endpoint"],
+            s3_parameters["access-key"],
+            s3_parameters["secret-key"],
+        )
+        if not success:
+            event.fail("Failed to upload metadata to provided S3")
+            return
+
+        # Run operations to prepare for the backup
+        success, error_message = self._pre_backup()
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+            return
+
+        # Perform the backup
+        success, error_message = self._backup(
+            s3_parameters["bucket"],
+            s3_directory,
+            s3_parameters["endpoint"],
+            s3_parameters["region"],
+            s3_parameters["access-key"],
+            s3_parameters["secret-key"],
+        )
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+
+            success, error_message = self._post_backup()
+            if not success:
+                logger.warning(error_message)
+
+            return
+
+        # Run operations to clean up after the backup
+        success, error_message = self._post_backup()
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+            return
+
+        event.set_results(
+            {
+                "backup-id": datetime_backup_requested,
+            }
+        )
+
+    def _retrieve_s3_parameters(self) -> Tuple[Dict, List[str]]:
+        """Retrieve S3 parameters from the S3 integrator relation."""
+        s3_parameters = self.s3_integrator.get_s3_connection_info()
+
+        required_parameters = [
+            "bucket",
+            "endpoint",
+            "region",
+            "path",
+            "access-key",
+            "secret-key",
+        ]
+        missing_parameters = [param for param in required_parameters if param not in s3_parameters]
+        if missing_parameters:
+            logger.warning(
+                f"Missing required S3 parameters in relation with S3 integrator: {missing_parameters}"
+            )
+
+        return s3_parameters, missing_parameters
+
+    def _can_unit_perform_backup(self) -> Tuple[bool, str]:
+        """Validates whether this unit can perform a backup."""
+        logger.info("Checking state and role of unit")
+
+        try:
+            state, role = self.charm._mysql.get_member_state()
+        except MySQLGetMemberStateError:
+            return False, "Error obtaining member state"
+
+        if role == "primary":
```

Comment:
clusters with planned_units=1 should be able to perform backups
```suggestion
        if role == "primary" and self.app.planned_units() > 1:
```

Reply:
addressed in [2f28871](https://github.com/canonical/mysql-k8s-operator/pull/136/commits/2f288712cb675a55668e3c75896aaba6feebe24d)